### ISSUE TITLE
Merge `twisted/internet/_utilspy3.py` into `twisted/internet/utils.py`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6240]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6240|
|Type|task|
|Created|2013-01-02 01:09:12Z|
|Branch|https://github.com/twisted/twisted/tree/internet-utils-mergs-6240-2|

Also merge the associated test module. See [#6183](https://github.com/twisted/twisted/issues/6183).

Attachments:

* [merge-py3utils.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a2/1a20cf59f0584ada3deeff6c1c5b4c11c691aec0/0f2a9a40ca9c53745cad5bd701cacfb2e0821899.patch) (13074 bytes) - added by dreamriver on 2013-03-04 02:40:24Z - 
* [merge-py3utils.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a2/1a20cf59f0584ada3deeff6c1c5b4c11c691aec0/669250416c2d956249054d93e8f4839c14b80103.patch) (13072 bytes) - added by dreamriver on 2013-03-04 03:35:54Z - NativeStringIO was having a method called on it, extra periods I didn't delete it
* [merge-py3utils-file-removed.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a2/1a20cf59f0584ada3deeff6c1c5b4c11c691aec0/93f32d5ec2917dc6188ea3248c49b271f4ef9ca3.patch) (18833 bytes) - added by dreamriver on 2013-03-04 05:26:40Z - files werent removed because i did rm instead of svn rm, fixed now
* [merge-py3utils-redux.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a2/1a20cf59f0584ada3deeff6c1c5b4c11c691aec0/ba69cd9ce41f8115735cba3d07130466e19edefc.patch) (19115 bytes) - added by genericnick on 2013-03-05 05:47:17Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6240 6240
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_internet_utils_mergs_6240_2 branches/internet-utils-mergs-6240-2
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1357088952000000 1357088952000000
changetime__1364599920000000 1364599920000000
version__None None
owner__tom_prince tom.prince

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="dreamriver's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|dreamriver commented|
|-|-|

i'm not sure why the tests didn't pick up the failure but patch2 is more correct. (is there a way to remove previously submitted patches?)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

(In [37385]) Branching to internet-utils-mergs-6240.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince set owner to genericnick|
|-|-|

Thanks for your contribution.

  1. You shouldn't try to port the rest of the module in this ticket, other than the minimal syntax changes to get the file to import.
    * That is, you should conventionalize the import of `StringIO` on `_PY3`, rather than changing the import.
  2. For the things that aren't ported, you need to make sure that they aren't accessible. See the `if _PY3:` blocks at the end of `twisted/python/util.py` and `twisted/python/test/test_util.py` to see how to do this.

Please resubmit for review after addressing the above points.

(Note for future reviewers: I've done the initial dance to restore history, as in [#6235](https://github.com/twisted/twisted/issues/6235) but haven't applied the above patch)

**Comment by twisted-trac** [comment]
|<img alt="genericnick's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|genericnick commented|
|-|-|

changes requested made

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to @tomprince<br>@tomprince set status to `assigned`|
|-|-|

  1. `reraise` needs to be imported unconditionally. (Apparently there isn't test coverage for this unfortunately)
  2. `admin/_twistedpython3.py` needs to be updated to include the merged files.
  3. (minor) Please keep the import like:
```
#!
try:
    import cStringIO as StringIO
except ImportError:
    import StringIO
```
   (as it was before merging) and don't do any import on `_PY3`.

I'll fix these up and submit for review. (I want somebody to review the svn merge dance as well).

**Comment by twisted-trac** [comment]
|<img alt="genericnick's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|genericnick commented|
|-|-|

I'm confused, can you explain why?

1. I looked through ```twisted.internet.utils``` and didn't find any hint of ```reraise``` why would I want to still import it in general?
2. I removed the _utilspy3 ones and didn't put the utils and test_iutils in. Per your last comment I shouldn't try and port the files as a whole, so I removed them from the ported list.
3. I tried to do just that. Unfortunately since I need to import ```NativeStringIO``` it isn't possible for me to keep ```StringIO.StringIO``` which is why I did it the way I did. Is there a better way?

Thanks for your patience and help.

Also, is there a way I should go about submitting additional patches when the first ones are inadequate? The way I have been doing it seems suboptimal.

**Comment by twisted-trac** [comment]
|<img alt="genericnick's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|genericnick commented|
|-|-|

First of all, let me say thanks for your patience and help. However, I'm confused, can you explain why?

1. I looked through ```twisted.internet.utils``` and didn't find any hint of ```reraise``` why would I want to still import it in general?

2. I removed the _utilspy3 ones and didn't put the utils and test_iutils in. Per your last comment I shouldn't try and port the files as a whole, so I removed them from the ported list.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince commented|
|-|-|

> 1. I looked through twisted.internet.utils and didn't find any hint of reraise why would I want to still import it in general?

https://twistedmatrix.com/trac/browser/branches/internet-utils-mergs-6240/twisted/internet/utils.py?rev=#L203

> 2. I removed the _utilspy3 ones and didn't put the utils and test_iutils in. Per your last comment I shouldn't try and port the files as a whole, so I removed them from the ported list.

Well, some of it was ported, so the modules should get installed, and the tests run, which is the main purpose of `admin/_twistedpython3.py`. The `del`s at the end of those modules will remove any trace of the unported parts, leaving just the ported parts useable and testable on python 3.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince commented|
|-|-|

> 3. I tried to do just that. Unfortunately since I need to import NativeStringIO it isn't possible for me to keep StringIO.StringIO which is why I did it the way I did. Is there a better way?

Since that code never actually gets run (just parsed), it doesn't in fact need `NativeStringIO`. It will when that code gets ported, and the change you did is probably the right thing to do there. (What happens is the function/classes get defined, then get deleted before they are ever called.)

>  Also, is there a way I should go about submitting additional patches when the first ones are inadequate? The way I have been doing it seems suboptimal.

Suboptimal how? If your patches have been applied to a branch, you should base future patches off the tip of the branch. That is often the case, but I didn't do that here. (If you have further patches for this ticket, they should be against the branch).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner<br>@tomprince set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @tomprince|
|-|-|

1. I don't think you need to do "del ProcessUtilsTests", the tests will skip properly.

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

(In [37873]) Branching to internet-utils-mergs-6240-2.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince commented|
|-|-|

> 1. I don't think you need to do "del ProcessUtilsTests", the tests will skip properly.

The code they are testing hasn't been ported yet. Deleting the tests is the idiom that has been used to indicate that. In any case, the process support in the reactor will get ported to python3 before the process support in t.i.utils, which will cause these tests to start failing.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_15"></a>|@tomprince set status to `closed`|
|-|-|

(In [37876]) Merge internet-utils-mergs-6240-2: Merge `twisted/internet/_utilspy3.py` into `twisted/internet/utils.py`.

Author: genericnick, tom.prince
Reviewers: tom.prince, therve
Fixes: [#6240](https://github.com/twisted/twisted/issues/6240)
